---
title: "Annotating variants with OpenCRAVAT, using Bioconductor"
author: Vince Carey^[stvjc@channing.harvard.edu]
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Annotating variants with OpenCRAVAT, using Bioconductor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: bcoc.bib
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
suppressPackageStartupMessages({
library(oc2bioc)
library(DT)
library(curatedTCGAData)
})
```

# Introduction to oc2bioc

The oc2bioc package (currently at
github.com/vjcitn/oc2bioc) incorporates the open-cravat python modules.
This workshop takes advantage of R functions and interfaces defined
in that package to illustrate the programmatic use of OpenCRAVAT
in R/Bioconductor.

Case studies of the use of the OpenCRAVAT GUI are published
in a Journal of Clinical Oncology article (@Pagel2020).

## Displaying and searching the resources

We acquire the metadata for all OpenCRAVAT 'modules' in real time
using `populate_module_set`.

```{r setup}
library(oc2bioc)
modset = populate_module_set()
modset
```
We transformed the module set object to a searchable HTML table:
```{r displ,echo=FALSE}
moddf = as.data.frame(modset)
moddf$description = gsub("In both the graphical interface..*", "", moddf$description)
MAXDESCLEN=85
lng = which(nchar(moddf$description)>MAXDESCLEN)
if (length(lng)>0) {
  moddf$description[lng] = paste0(substr(moddf$description[lng],1,MAXDESCLEN), "...")
  }
suppressPackageStartupMessages(suppressMessages(library(DT)))
datatable(moddf, options=list(lengthMenu=c(5,10,25,100)))
```

## Single-variant queries

The `queryOC` function uses the OpenCRAVAT REST API to acquire
information on a single position in GRCh38 coordinates.  This
function requires that a registered username and password
be supplied.  Visit `run.opencravat.org` to set a username
and password.

An example API response is available in `var_in_tx` in the
`oc2bioc` package.  This was returned by the call
`queryOC(chr="chr7", pos="140753336", annotators=c("pubmed", "segway_breast", "chasmplus_BRCA"))`
```{r lkex}
var_in_tx
```
We use `httr::content` to explore this result.
```{r lkc}
vcon = httr::content(var_in_tx)
names(vcon)
```
A peek at the result:
```{r dolv}
names(vcon)
str(head(vcon,3))
```

Click [here](`r vcon$pubmed$term`) for the
PubMed references related to the gene harboring the
query variant.

The `crx` component of the response provides information
on variant impacts at the transcript level.
```{r lkcrx}
nl = rjson::fromJSON(vcon$crx$all_mappings)
DT::datatable(data.frame(do.call(rbind, unlist(nl, recursive=FALSE))))
```

## Viewing and assembling collections of variants for annotation

We'll use TCGA as a source of realistic variant sets.
The curatedTCGAData package will be used.

### Acquiring variants and filtering to SNVs

```{r docur}
library(curatedTCGAData)
suppressMessages({
acc = curatedTCGAData("ACC", assays="Mutation", dry.run=FALSE,
   cache=BiocFileCache::bfccache(BiocFileCache::BiocFileCache(ask=FALSE)),
   version="2.0.1")
})
eacc = experiments(acc)[[1]]
eacc
muts = as(eacc, "GRangesList")
sum(elementNROWS(muts))
DT::datatable(as.data.frame(head(muts[[1]][,1:4])))
```
Notice that there is a variant with width 24.  We will
focus on single-nucleotide variants (SNVs).
```{r dosn}
snvs = unlist(muts[width(muts)==1])
sum(elementNROWS(snvs))
```

### Visualizing variants in the context of gene models

We will use a function based on
the TnT package to visualize variant locations
in the context of gene-like features.

Here's a 1Mb slice:
```{r dodod}
TnTdemo(acc, viewstart=6.4e7, viewend=6.5e7)
```

Now we drill down a bit, to a 0.1Mb region near MAP4K2:
```{r dododo}
TnTdemo(acc, viewstart=6.455e7, viewend=6.465e7)
```

## Annotating a large collection of variants

We use the function `make_oc_POSTable` to transform
a GRanges instance into a data.frame that can be
saved to a file and submitted to the OpenCRAVAT API.

```{r lksnpo}
mdf = make_oc_POSTable(snvs)
head(mdf)
```

The data.frame must be formatted for ingestion by OpenCRAVAT.
The following call to write.table will accomplish this.
```{r dor}
tf = tempfile()
write.table(mdf, file=tf, sep="\t", col.names=FALSE, 
   row.names=FALSE, quote=FALSE)
head(read.delim(tf, h=FALSE, sep="\t"))
```




# References
